src/core/qgstemporalnavigationobject.cpp

```diff
@@ -0,0 +1,175 @@
+/***************************************************************************
+                         qgstemporalnavigationobject.cpp
+                         ---------------
+    begin                : March 2020
+    copyright            : (C) 2020 by Samweli Mwakisambwe
+    email                : samweli at kartoza dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgstemporalnavigationobject.h"
+#include "qgis.h"
+
+QgsTemporalNavigationObject::QgsTemporalNavigationObject( QObject *parent )
+  : QObject( parent )
+{
+  mNewFrameTimer = new QTimer( this );
+
+  connect( mNewFrameTimer, &QTimer::timeout,
+           this,  qgis::overload<>::of( &QgsTemporalNavigationObject::timerTimeout ) );
+}
+
+
+void QgsTemporalNavigationObject::timerTimeout()
+{
+  if ( mPlayBackMode == PlaybackMode::Forward )
+    forward();
+  if ( mPlayBackMode == PlaybackMode::Reverse )
+    backward();
+  return;
+}
+
+QgsDateTimeRange QgsTemporalNavigationObject::dateTimeRangeForFrameNumber( long long frame ) const
+{
+  QDateTime start = mTemporalExtents.begin();
+
+  long long pastFrame = frame - 1;
+  if ( pastFrame < 0 )
+    pastFrame = 0;
+
+  QDateTime begin = start.addSecs( pastFrame * mFrameDuration.seconds() );
+  QDateTime end = start.addSecs( frame * mFrameDuration.seconds() );
+
+  if ( end <= mTemporalExtents.end() )
+    return QgsDateTimeRange( begin, end );
+
+  return mTemporalExtents;
+}
+
+void QgsTemporalNavigationObject::setTemporalExtents( QgsDateTimeRange temporalExtents )
+{
+  mTemporalExtents = temporalExtents;
+}
+
+QgsDateTimeRange QgsTemporalNavigationObject::temporalExtents() const
+{
+  return mTemporalExtents;
+}
+
+void QgsTemporalNavigationObject::setCurrentFrameNumber( long long frameNumber )
+{
+  if ( frameNumber < 0 ||
+       frameNumber >= totalFrameCount() )
+  {
+    pause();
+    return;
+  }
+
+  if ( mCurrentFrameNumber != frameNumber )
+  {
+    mCurrentFrameNumber = frameNumber;
+    QgsDateTimeRange range = dateTimeRangeForFrameNumber( mCurrentFrameNumber );
+    emit updateTemporalRange( range );
+  }
+}
+
+long long QgsTemporalNavigationObject::currentFrameNumber() const
+{
+  return mCurrentFrameNumber;
+}
+
+void QgsTemporalNavigationObject::setFrameDuration( QgsInterval frameDuration )
+{
+  mFrameDuration = frameDuration;
+}
```

Comment:
again, we probably should rewind when this is called? otherwise we'd need to resync the current frame number based on the current time range using the old frameDuration and convert it to the new frame number using the new frameDuration